### ISSUE TITLE
ValueError: no signature found for builtin <method 'mro' of 'type' objects>

### ISSUE THREAD

**Issue by SunDwarf** [agronholm/sphinx-autodoc-typehints#18]
```
»»»» python -V
Python 3.6.1
»»»» cat /tmp/sphinx-err-0dalwnxn.log
# Sphinx version: 1.6.3
# Python version: 3.6.1 (CPython)
# Docutils version: 0.13.1 release
# Jinja2 version: 2.9.6
# Last messages:
#   building [html]: targets for 4 source files that are out of date
#   
#   updating environment:
#   
#   4 added, 0 changed, 0 removed
#   
#   reading sources... [ 25%] hilvl/connecting
#   
#   reading sources... [ 50%] hilvl/tables
#   
# Loaded extensions:
#   alabaster (0.7.10) from /home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/alabaster/__init__.py
#   sphinx.ext.autodoc (1.6.3) from /home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/ext/autodoc.py
#   sphinx.ext.doctest (1.6.3) from /home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/ext/doctest.py
#   sphinx.ext.intersphinx (1.6.3) from /home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/ext/intersphinx.py
#   sphinx.ext.coverage (1.6.3) from /home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/ext/coverage.py
#   sphinx.ext.ifconfig (1.6.3) from /home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/ext/ifconfig.py
#   sphinx.ext.viewcode (1.6.3) from /home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/ext/viewcode.py
#   sphinxcontrib.asyncio (1.0) from /home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinxcontrib/asyncio.py
#   sphinx_autodoc_typehints (unknown version) from /home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx_autodoc_typehints.py
Traceback (most recent call last):
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/cmdline.py", line 306, in main
    app.build(opts.force_all, filenames)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/application.py", line 339, in build
    self.builder.build_update()
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/builders/__init__.py", line 331, in build_update
    'out of date' % len(to_build))
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/builders/__init__.py", line 344, in build
    updated_docnames = set(self.env.update(self.config, self.srcdir, self.doctreedir))
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/environment/__init__.py", line 583, in update
    self._read_serial(docnames, self.app)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/environment/__init__.py", line 602, in _read_serial
    self.read_doc(docname, app)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/environment/__init__.py", line 723, in read_doc
    pub.publish()
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/docutils/core.py", line 217, in publish
    self.settings)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/io.py", line 74, in read
    self.parse()
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/docutils/readers/__init__.py", line 78, in parse
    self.parser.parse(self.input, document)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/docutils/parsers/rst/__init__.py", line 185, in parse
    self.statemachine.run(inputlines, document, inliner=self.inliner)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/docutils/parsers/rst/states.py", line 170, in run
    input_source=document['source'])
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/docutils/statemachine.py", line 239, in run
    context, state, transitions)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/docutils/statemachine.py", line 460, in check_line
    return method(match, context, next_state)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/docutils/parsers/rst/states.py", line 2745, in underline
    self.section(title, source, style, lineno - 1, messages)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/docutils/parsers/rst/states.py", line 326, in section
    self.new_subsection(title, lineno, messages)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/docutils/parsers/rst/states.py", line 394, in new_subsection
    node=section_node, match_titles=True)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/docutils/parsers/rst/states.py", line 281, in nested_parse
    node=node, match_titles=match_titles)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/docutils/parsers/rst/states.py", line 195, in run
    results = StateMachineWS.run(self, input_lines, input_offset)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/docutils/statemachine.py", line 239, in run
    context, state, transitions)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/docutils/statemachine.py", line 460, in check_line
    return method(match, context, next_state)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/docutils/parsers/rst/states.py", line 2745, in underline
    self.section(title, source, style, lineno - 1, messages)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/docutils/parsers/rst/states.py", line 326, in section
    self.new_subsection(title, lineno, messages)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/docutils/parsers/rst/states.py", line 394, in new_subsection
    node=section_node, match_titles=True)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/docutils/parsers/rst/states.py", line 281, in nested_parse
    node=node, match_titles=match_titles)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/docutils/parsers/rst/states.py", line 195, in run
    results = StateMachineWS.run(self, input_lines, input_offset)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/docutils/statemachine.py", line 239, in run
    context, state, transitions)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/docutils/statemachine.py", line 460, in check_line
    return method(match, context, next_state)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/docutils/parsers/rst/states.py", line 2318, in explicit_markup
    nodelist, blank_finish = self.explicit_construct(match)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/docutils/parsers/rst/states.py", line 2330, in explicit_construct
    return method(self, expmatch)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/docutils/parsers/rst/states.py", line 2073, in directive
    directive_class, match, type_name, option_presets)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/docutils/parsers/rst/states.py", line 2122, in run_directive
    result = directive_instance.run()
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/ext/autodoc.py", line 1827, in run
    documenter.generate(more_content=self.content)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/ext/autodoc.py", line 1127, in generate
    self.document_members(all_members)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/ext/autodoc.py", line 1524, in document_members
    ModuleLevelDocumenter.document_members(self, all_members)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/ext/autodoc.py", line 1048, in document_members
    check_module=members_check_module and not isattr)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/ext/autodoc.py", line 1114, in generate
    sig = self.format_signature()
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/ext/autodoc.py", line 1320, in format_signature
    return Documenter.format_signature(self)  # type: ignore
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/ext/autodoc.py", line 758, in format_signature
    self.object, self.options, args, retann)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/application.py", line 493, in emit_firstresult
    return self.events.emit_firstresult(event, self, *args)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/events.py", line 84, in emit_firstresult
    for result in self.emit(name, *args):
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/events.py", line 79, in emit
    results.append(callback(*args))
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx_autodoc_typehints.py", line 118, in process_signature
    argspec = getargspec(obj)
  File "/home/laura/.local/share/virtualenvs/katagawa/lib/python3.6/site-packages/sphinx/util/inspect.py", line 52, in getargspec
    sig = inspect.signature(func)
  File "/usr/lib64/python3.6/inspect.py", line 3001, in signature
    return Signature.from_callable(obj, follow_wrapped=follow_wrapped)
  File "/usr/lib64/python3.6/inspect.py", line 2751, in from_callable
    follow_wrapper_chains=follow_wrapped)
  File "/usr/lib64/python3.6/inspect.py", line 2230, in _signature_from_callable
    skip_bound_arg=skip_bound_arg)
  File "/usr/lib64/python3.6/inspect.py", line 2060, in _signature_from_builtin
    raise ValueError("no signature found for builtin {!r}".format(func))
ValueError: no signature found for builtin <method 'mro' of 'type' objects>
```

**Comment by 132nd-etcher** [comment]
I think I've hit the same issue using PyQt5:
```
  File "d:\anaconda3\envs\emft\lib\site-packages\docutils\parsers\rst\states.py", line 2318, in explicit_markup
    nodelist, blank_finish = self.explicit_construct(match)
  File "d:\anaconda3\envs\emft\lib\site-packages\docutils\parsers\rst\states.py", line 2330, in explicit_construct
    return method(self, expmatch)
  File "d:\anaconda3\envs\emft\lib\site-packages\docutils\parsers\rst\states.py", line 2073, in directive
    directive_class, match, type_name, option_presets)
  File "d:\anaconda3\envs\emft\lib\site-packages\docutils\parsers\rst\states.py", line 2122, in run_directive
    result = directive_instance.run()
  File "d:\anaconda3\envs\emft\lib\site-packages\sphinx\ext\autodoc.py", line 1827, in run
    documenter.generate(more_content=self.content)
  File "d:\anaconda3\envs\emft\lib\site-packages\sphinx\ext\autodoc.py", line 1127, in generate
    self.document_members(all_members)
  File "d:\anaconda3\envs\emft\lib\site-packages\sphinx\ext\autodoc.py", line 1048, in document_members
    check_module=members_check_module and not isattr)
  File "d:\anaconda3\envs\emft\lib\site-packages\sphinx\ext\autodoc.py", line 1127, in generate
    self.document_members(all_members)
  File "d:\anaconda3\envs\emft\lib\site-packages\sphinx\ext\autodoc.py", line 1524, in document_members
    ModuleLevelDocumenter.document_members(self, all_members)
  File "d:\anaconda3\envs\emft\lib\site-packages\sphinx\ext\autodoc.py", line 1048, in document_members
    check_module=members_check_module and not isattr)
  File "d:\anaconda3\envs\emft\lib\site-packages\sphinx\ext\autodoc.py", line 1114, in generate
    sig = self.format_signature()
  File "d:\anaconda3\envs\emft\lib\site-packages\sphinx\ext\autodoc.py", line 1339, in format_signature
    return Documenter.format_signature(self)  # type: ignore
  File "d:\anaconda3\envs\emft\lib\site-packages\sphinx\ext\autodoc.py", line 758, in format_signature
    self.object, self.options, args, retann)
  File "d:\anaconda3\envs\emft\lib\site-packages\sphinx\application.py", line 493, in emit_firstresult
    return self.events.emit_firstresult(event, self, *args)
  File "d:\anaconda3\envs\emft\lib\site-packages\sphinx\events.py", line 84, in emit_firstresult
    for result in self.emit(name, *args):
  File "d:\anaconda3\envs\emft\lib\site-packages\sphinx\events.py", line 79, in emit
    results.append(callback(*args))
  File "d:\anaconda3\envs\emft\lib\site-packages\sphinx_autodoc_typehints.py", line 118, in process_signature
    argspec = getargspec(obj)
  File "d:\anaconda3\envs\emft\lib\site-packages\sphinx\util\inspect.py", line 52, in getargspec
    sig = inspect.signature(func)
  File "d:\anaconda3\envs\emft\lib\inspect.py", line 3001, in signature
    return Signature.from_callable(obj, follow_wrapped=follow_wrapped)
  File "d:\anaconda3\envs\emft\lib\inspect.py", line 2751, in from_callable
    follow_wrapper_chains=follow_wrapped)
  File "d:\anaconda3\envs\emft\lib\inspect.py", line 2230, in _signature_from_callable
    skip_bound_arg=skip_bound_arg)
  File "d:\anaconda3\envs\emft\lib\inspect.py", line 2060, in _signature_from_builtin
    raise ValueError("no signature found for builtin {!r}".format(func))
ValueError: no signature found for builtin <unbound PYQT_SIGNAL threading_signal[object, object, object, object]>```

**Comment by agronholm** [comment]
@132nd-etcher what did the code look like that produced this error?

**Comment by 132nd-etcher** [comment]
I've been trying, but so far I've been unable to pinpoint the location in source that triggers the issue. I'm still searching for it =)

I'm pretty new to Sphinx, maybe you have a pointer for me ?

**Comment by agronholm** [comment]
This is easy. Modify the sphinx_autodoc_typehints code to catch the ValueError on line 119 and in the error handler, print the function object.

**Comment by agronholm** [comment]
If that doesn't give you enough information, you may want to print the module name of the function and line numbers of its code object.

**Comment by 132nd-etcher** [comment]
I found the culprit:
`threading_signal = pyqtSignal(object, object, object, object, name='threading_signal')`

Removing the fanciness with `threading_signal = pyqtSignal()`raises the same error: 
```
Exception occurred:
  File "d:\anaconda3\envs\emft\lib\inspect.py", line 2060, in _signature_from_builtin
    raise ValueError("no signature found for builtin {!r}".format(func))
ValueError: no signature found for builtin <unbound PYQT_SIGNAL MainUiThreading.threading_signal[]>
```
Now I'm completely out of my league here, but since PyQt is a wrapper to Qt, which is written (I guess?) in C, maybe that's the issue ?

**Comment by 132nd-etcher** [comment]
Should I try subclassing pyqtSignal to try and fool the typing ?

**Comment by 132nd-etcher** [comment]
Searching for declaration of pyqtSignal leads to:
```
    def __init__(self, *args, **kwargs): # real signature unknown
        pass
```

It makes sense autodoc-typehints gets confused I guess.

**Comment by agronholm** [comment]
Ok, so this isn't even a function we're dealing with. I was just hoping to be able to write a test for this but I won't add pyqt as a test dependency for this. Maybe @SunDwarf could shed some light on the issue.

**Comment by agronholm** [comment]
Meanwhile I'm going to commit an easy fix for this: catch the `ValueError` and return early, as is done with `TypeError`.

**Comment by 132nd-etcher** [comment]
Is there any way to setup autodoc-typehint to ignore the issue and carry on on its merry way with building the docs, omitting this weirdo ?

**Comment by agronholm** [comment]
No, you'll need to apply the quick fix or wait until I release v1.2.1.

**Comment by 132nd-etcher** [comment]
Happy to ! Thx for building this, man =)

**Comment by 132nd-etcher** [comment]
Tested, it works for me !

**Comment by agronholm** [comment]
@132nd-etcher I've now released v1.2.1.
